python/setup.py

```diff
@@ -82,18 +95,113 @@
     "tune": ["tabulate"],
 }
 
+def is_native_windows_or_msys():
+    """Check to see if we are running on native Windows, but NOT WSL (which is seen as Linux)."""
+    return sys.platform == 'msys' or sys.platform == 'win32'
 
-class build_ext(_build_ext.build_ext):
-    def run(self):
+def is_invalid_windows_platform():
+    # 'GCC' check is how you detect MinGW: https://github.com/msys2/MINGW-packages/blob/abd06ca92d876b9db05dd65f27d71c4ebe2673a9/mingw-w64-python2/0410-MINGW-build-extensions-with-GCC.patch#L53
+    return sys.platform == 'msys' or sys.platform == 'win32' and sys.version_string and 'GCC' in sys.version_string
+
+def makedirs(path, exist_ok):
+    """Python 2-compatible version of makedirs().
+
+    WARNING: path components to create should NOT include ".." components
+    See documentation: https://docs.python.org/3/library/os.html#os.makedirs
+    """
+    if sys.version_info[:2] >= (3, 2):
+        os.makedirs(path, exist_ok=exist_ok)
+    else:
+        try:
+            os.makedirs(path)
+        except OSError as ex:
+            if not (exist_ok and ex.errno == errno.EEXIST and os.path.isdir(path)):
+                raise
+
+def zip_extract_members(zip_file, desired_members, target_dir):
```

Comment:
It would really be great to get rid of this function. This functionality is alreayd implement in the `zip` module. Why don't we use it instead of reimplementing it?

Reply:
I'm not reimplementing anything -- the functionality doesn't exist in the ZIP module unfortunately. It only allows extracting the entire archive or a single file, not just a subdirectory.

Reply:
Can we use the file `https://files.pythonhosted.org/packages/cd/5a/cbdf36134804809d55ffd4c248343bd36680a92b6425885a3fd204d32f7b/pickle5-0.0.9.tar.gz` which doesn't have subdirectories?

Reply:
I think the contents aren't the same? But also it seems to me there's still an extra subdirectory like before (as is conventional with tar files)?

Reply:
```
In [1]: import tarfile 
   ...: tf = tarfile.open("pickle5-0.0.9.tar.gz")                                                                                                                                    

In [2]: tf.extractall()
```

works for me

Reply:
Are you sure? I'm confused... it seems to create the same directory structure as
```
import zipfile
>>> zf = zipfile.ZipFile("pickle5-backport-5186f9ca4ce55ae530027db196da51e08208a16b.zip", "r")
>>> zf.extractall()
```
In your case I get a `pickle5-0.0.9` folder, in this case I get a `pickle5-backport-5186f9ca4ce55ae530027db196da51e08208a16b` folder, both of which contain `setup.py`.

Reply:
Looks like that's indeed the case, but `pickle5-0.0.9` is pretty short, so the danger of path lenght overlow is pretty small and we care less about creating an additional subdirectory :)

Reply:
Okay sounds good! Will modify it in a bit.